File: src/cxx-dev/LiENa/lienaStructure/lienaDatagram/lienaDatagram.cpp
#include "lienaDatagram.h"


/**
 * @brief lienaDatagram::lienaDatagram
 */
lienaDatagram::lienaDatagram(uint32_t size, char** chunk){
    this->size = size;
    this->byteArray = *chunk;
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::set_size
//! \param size
//!
void lienaDatagram::setSize(int size){
    this->size = size;
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getDataType
//! \return
//!
uint64_t lienaDatagram::getMessageID(){
    return (uint64_t(byteArray[0]))*256*256*256*256*256*256*256  +
           (uint64_t(byteArray[1]))*256*256*256*256*256*256  +
           (uint64_t(byteArray[2]))*256*256*256*256*256  +
           (uint64_t(byteArray[3]))*256*256*256*256 +
           (uint64_t(byteArray[4]))*256*256*256 +
           (uint64_t(byteArray[5]))*256*256 +
           (uint64_t(byteArray[6]))*256 +
           (uint64_t(byteArray[7]));
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getOrigineId
//! \return
//!
uint32_t lienaDatagram::getOriginId(){
    return  (uint32_t(byteArray[12]))*256*256*256  +
            (uint8_t(byteArray[13]))*256*256  +
            (uint8_t(byteArray[14]))*256 +
            (uint8_t(byteArray[15]));
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::writeValueInFiveBytes
//! \param start
//! \param value
//!
void lienaDatagram::writeValueInEightBytes(int start, uint64_t value){
    byteArray[start]     = (value&0xff00000000000000)>>56;
    byteArray[start + 1] = (value&0x00ff000000000000)>>48;
    byteArray[start + 2] = (value&0x0000ff0000000000)>>40;
    byteArray[start + 3] = (value&0x000000ff00000000)>>32;
    byteArray[start + 4] = (value&0x00000000ff000000)>>24;
    byteArray[start + 5] = (value&0x0000000000ff0000)>>16;
    byteArray[start + 6] = (value&0x000000000000ff00)>>8;
    byteArray[start + 7] = (value&0x00000000000000ff);
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getDeviceClass
//! \return
//!
uint8_t lienaDatagram::getDeviceClass(){
    return uint8_t(this->getOriginId() / (256*256*256*32));
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getManufacture
//! \return
//!
uint32_t lienaDatagram::getManufacture(){
    return uint32_t((this->getOriginId() % (256*256*256*32)) / (256*128));
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getDeviceType
//! \return
//!
uint8_t lienaDatagram::getDeviceType(){
    return uint8_t((this->getOriginId() % (256*128)) / (256*2));
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getDeviceVision
//! \return
//!
uint8_t lienaDatagram::getDeviceVision(){
    return uint8_t((this->getOriginId() %  (256*2)) /8);
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getDeviceIndex
//! \return
//!
uint8_t lienaDatagram::getDeviceIndex(){
    return uint8_t(this->getOriginId() % 8);

}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getTargetId
//! \return
//!
uint32_t lienaDatagram::getTargetId(){
    return (uint32_t(byteArray[8]))*256*256*256 +
           (uint32_t(byteArray[9]))*256*256 +
           (uint32_t(byteArray[10]))*256 +
           (uint32_t(byteArray[11]));
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getTimeStampes
//! \return
//!
uint32_t lienaDatagram::getTimeStampes(){
    return uint8_t(byteArray[16])*256*256*256*256*256*256*256  +
           uint8_t(byteArray[17])*256*256*256*256*256*256 +
           uint8_t(byteArray[18])*256*256*256*256*256 +
           uint8_t(byteArray[19])*256*256*256*256 +
           uint8_t(byteArray[20])*256*256*256 +
           uint8_t(byteArray[21])*256*256 +
           uint8_t(byteArray[22])*256 +
           uint8_t(byteArray[23]);
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getDLC
//! \return
//!
uint32_t lienaDatagram::getDLC(){
    return uint8_t(byteArray[24])*256*256*256 +
           uint8_t(byteArray[25])*256*256 +
           uint8_t(byteArray[26])*256 +
           uint8_t(byteArray[27]);
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getIgtDatagramBody
//! \return
//!
char* lienaDatagram::getBody(){
    char *body = new char[size-HEADER_SIZE];
    strncpy(body, this->byteArray + HEADER_SIZE, this->size - HEADER_SIZE - 1);
    return body;
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaDatagram::getByteArray
//! \return
//!
char** lienaDatagram::getByteArray(){
    return &this->byteArray;
}

//! -----------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtDatagram::print
//!
QString lienaDatagram::print(){
    QString print_str;
    for(int i = 0; i < 26; i++){
        print_str+= " " + QString::number((uint8_t)this->byteArray[i]);
    }
    return print_str;
}

File: src/py-dev/LiENa/LiENaStructure/LiENaDatagram/LienaDatagram.py
from LiENaBasic.lienaDefinition import *


class LienaDatagram:
    def __init__(self, size, byte_array):
        self.size = size
        self.byteArray = byte_array
        self.addr = []
        self.port = 0
        self.body = ''
        self.index = 0

    def set_size(self, size):
        self.size = size

    def write_value_in_eight_byte(self, start, value):
        self.byteArray[start]     = (value & 0xff00000000000000) >> 56
        self.byteArray[start + 1] = (value & 0x00ff000000000000) >> 48
        self.byteArray[start + 2] = (value & 0x0000ff0000000000) >> 40
        self.byteArray[start + 3] = (value & 0x000000ff00000000) >> 32
        self.byteArray[start + 4] = (value & 0x00000000ff000000) >> 24
        self.byteArray[start + 5] = (value & 0x0000000000ff0000) >> 16
        self.byteArray[start + 6] = (value & 0x000000000000ff00) >> 8
        self.byteArray[start + 7] = (value & 0x00000000000000ff)

    def get_byte_array(self):
        return self.byteArray

    def get_message_id(self):
        return self.byteArray[0] * (256 ** 7) \
               + self.byteArray[1] * (256 ** 6) \
               + self.byteArray[2] * (256 ** 5) \
               + self.byteArray[3] * (256 ** 4) \
               + self.byteArray[4] * (256 ** 3) \
               + self.byteArray[5] * (256 ** 2) \
               + self.byteArray[6] * 256 \
               + self.byteArray[7]

    def get_origin_id(self):
        return self.byteArray[12] * 256 ** 3 + self.byteArray[13] * 256 ** 2 + self.byteArray[14] * 256 + self.byteArray[15]

    def get_device_class(self):
        return self.get_origin_id() / (32 * 256 ** 3)

    def get_manufacture(self):
        return self.get_origin_id() % (32 * 256 ** 3) / (256 * 2)

    def get_device_type(self):
        return self.get_origin_id() % (256 ** 2) / (256 * 4)

    def get_device_vision(self):
        return self.get_origin_id() % (256 * 4) / 8

    def get_device_index(self):
        return self.get_origin_id() % 8

    def get_target_id(self):
        return self.byteArray[8] * 256 ** 3 + self.byteArray[9] * 256 ** 2 + self.byteArray[10] * 256 + self.byteArray[11]

    def get_time_stamps(self):
        return   self.byteArray[16] * (256 ** 7) \
               + self.byteArray[17] * (256 ** 6) \
               + self.byteArray[18] * (256 ** 5) \
               + self.byteArray[19] * (256 ** 4) \
               + self.byteArray[20] * (256 ** 3) \
               + self.byteArray[21] * (256 ** 2) \
               + self.byteArray[22] * (256 ** 1)  \
               + self.byteArray[23] * (256 ** 0)

    def get_dlc(self):
        return self.byteArray[24] * 256 ** 3 + self.byteArray[25] * 256 ** 2 + self.byteArray[26] * 256 + self.byteArray[27]

    def get_body(self):
        self.body = self.byteArray[HEAD_SIZE:self.size]
        return self.body

    def get_index(self):
        self.index = self.byteArray[19]
        return self.index

    def print_message(self):
        msg = ""
        message_len = len(self.byteArray)
        for i in range(30):
            msg += str(self.byteArray[i]) + "; "
        return msg
